### PR TITLE
Test isnan

### PR DESCRIPTION
**Pull request by dneto0**
<hr>

**Author checklist for test code/plans:**

- [x] All outstanding work is tracked with "TODO" in a test/file description or `.unimplemented()` on a test.
- [x] New helpers, if any, are documented using `/** doc comments */` and can be found via `helper_index.txt`.
- [x] (Optional, sometimes not possible.) Tests pass (or partially pass without unexpected issues) in an implementation. (Add any extra details above.)

**[Reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) for test code/plans:** (Note: feel free to pull in other reviewers at any time for any reason.)

- [ ] The test path is reasonable, the [description](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) "describes the test, succinctly, but in enough detail that a reader can read only the test plans in a file or directory and evaluate the completeness of the test coverage."
- [ ] Tests appear to cover this area completely, except for outstanding TODOs. Validation tests use control cases.
    (This is critical for coverage. Assume anything without a TODO will be forgotten about forever.)
- [ ] Existing (or new) test helpers are used where they would reduce complexity.
- [ ] TypeScript code is readable and understandable (is unobtrusive, has reasonable type-safety/verbosity/dynamicity).

### CODE REVIEW

File: src/webgpu/shader/execution/builtin/isnan.spec.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      const expected_bits = NumberRepr.fromF32(expected ? 1.0 : 0).bits[0];
+      return {
+        input: NumberRepr.fromU32(bits),
+        expected: [NumberRepr.fromU32(expected_bits)],
+      };
+    };
+
+    const casesFromUints = [
+      makeUint(0x7f80_0000, false), // infinity
+      makeUint(0xff80_0000, false), // -infinity
```
Comment (sarahM0) [inline src/webgpu/shader/execution/builtin/isnan.spec.ts:81]: I'd suggest using the constants form https://github.com/gpuweb/cts/blob/2961e70ac940aa861f924efd1ec97405a0afc4e5/src/webgpu/shader/execution/builtin/builtin.ts#L211.

File: src/webgpu/shader/execution/builtin/isinf.spec.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      // Non-infinity
+      make(0.0, false),
+      make(10.0, false),
+      make(-10.0, false),
+      // Infinities
+      make(Infinity, true),
+      make(-Infinity, true),
+      // NaNs
+      make(NaN, false),
+      make(-NaN, false),
```
Comment (sarahM0) [inline src/webgpu/shader/execution/builtin/isinf.spec.ts:51]: I'd suggest using https://github.com/gpuweb/cts/blob/2961e70ac940aa861f924efd1ec97405a0afc4e5/src/webgpu/shader/execution/builtin/builtin.ts#L201

File: src/webgpu/shader/values.ts [hunk 3]
```diff
@@ -63,3 +63,29 @@ export function normalF32Examples(): Array<number> {
...
+    const bits = sign_and_exponent | 0x40_0000;
+    // Only the most significant bit of the mantissa is set.
+    result.push(bits);
+
+    // Quiet and signalling NaNs differ based on the most significant bit
+    // of the mantissa. Try both.
+    for (const quiet_signalling in [0, 0x40_0000]) {
+      // Set each of the lower bits.
+      for (let lower_bits = 1; lower_bits < 0x40_0000; lower_bits <<= 1) {
+        const bits = sign_and_exponent | ((quiet_signalling as unknown) as number) | lower_bits;
```
Comment (ben-clayton) [inline src/webgpu/shader/values.ts:85]: ditto

File: src/webgpu/shader/values.ts [hunk 4]
```diff
@@ -63,3 +63,29 @@ export function normalF32Examples(): Array<number> {
...
 }
+
+/** Returns an array of 32-bit NaNs, as Uint32 bit patterns.
+ * NaNs have: maximum exponent, but the mantissa is not zero.
+ */
+export function nanF32BitsExamples(): Array<number> {
+  const result: number[] = [];
+  const exponent_bit = 0x7f80_0000;
+  const sign_bits: [number, number] = [0, 0x8000_0000];
+  for (const sign_bit in sign_bits) {
```
Comment (ben-clayton) [inline src/webgpu/shader/values.ts:74]: I think you want `for (const sign_bit of sign_bits) {`

`x in y` [gives you the index of each element](https://www.tutorialsteacher.com/typescript/for-loop), and my IDE claims the type of `sign_bit` is a `string` (?!). I'm guessing this is why on L75 you're doing weird casting.

I see you're doing the same thing above, which I think might be buggy. Sorry I didn't spot this before.

**Review by ben-clayton — CHANGES_REQUESTED** [review]

File: src/webgpu/shader/values.ts (hunk 4 above)
Comment (dneto0) [reply src/webgpu/shader/values.ts:74]: Yikes. You're right.  Yeah, I was seeing that 'strings' stuff too.

> The for...in statement iterates over all enumerable properties of an object that are keyed by strings ....

File: src/webgpu/shader/execution/builtin/isnan.spec.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      .combine('vectorize', [undefined, 2, 3, 4] as const)
+  )
+  .fn(async t => {
+    const cases = [
+      // Simple non-NaN
+      { input: f32(0.0), expected: False },
+      { input: f32(10.0), expected: False },
+      { input: f32(-10.0), expected: False },
+      // Infinities
+      { input: f32(Infinity), expected: False },
```
Comment (ben-clayton) [inline src/webgpu/shader/execution/builtin/isnan.spec.ts:45]: I'm really sure why we have two sets of tests for infinity - the shader will see the same thing for both approaches. I'd suggest those in `kBit.f32.infinity`

**Review by ben-clayton — APPROVED** [review]
LGTM, although I'm confused by the dual infinity cases

File: src/webgpu/shader/execution/builtin/isnan.spec.ts (hunk 5 above)
Comment (dneto0) [reply src/webgpu/shader/execution/builtin/isnan.spec.ts:45]: I wanted to see idiomatic JavaScript infinities.  
But I can take it out.

File: src/webgpu/shader/execution/builtin/isnan.spec.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+`
+  )
+  // https://gpuweb.github.io/gpuweb/wgsl/#floating-point-evaluation says:
+  //
+  //    Implementations may assume that NaNs, inifinities are not present
+  //
+  //       Note: This means some functions (e.g. isInf, isNan, min and max) may not return
+  //       the expected result due to optimizations about the presence of NaNs and infinities.
+  //
+  // TODO: For now, test assuming NaNs are handled strictly, according to IEEE-754. Relax
```
Comment (dneto0) [inline src/webgpu/shader/execution/builtin/isnan.spec.ts:31]: Hey @alan-baker please confirm you want me to allow any boolean answer for the NaN cases, even with this TODO.

File: src/webgpu/shader/execution/builtin/isnan.spec.ts (hunk 5 above)
Comment (ben-clayton) [reply src/webgpu/shader/execution/builtin/isnan.spec.ts:45]: These should end up as the same bit representation by the time it reaches the shader, so you're really testing the test framework, not the built-in. We should be testing this - but this isn't the right place for it.

File: src/webgpu/shader/execution/builtin/isnan.spec.ts (hunk 6 above)
Comment (dneto0) [reply src/webgpu/shader/execution/builtin/isnan.spec.ts]: Changed to allow either answer.
Will file a post-v1 request for float-controls feature to allow preservation of infinities and nans at least.

**Review by alan-baker — APPROVED** [review]

**Pull request merged**